File: frontend/src/lib/scoring.ts
// Score computation — mirrors backend utils/scoring.py

export type UoMType = 'min' | 'max' | 'timeline' | 'zero'

export function computeScore(
  uomType: UoMType,
  target: number,
  achievement: number,
  completionDate?: string | null,
  deadline?: string | null,
): number {
  switch (uomType) {
    case 'min':
      // Higher is better (sales revenue)
      if (target === 0) return achievement >= 0 ? 1.0 : 0.0
      return Math.round((achievement / target) * 10000) / 10000

    case 'max':
      // Lower is better (TAT, cost)
      if (achievement === 0) return 1.0
      if (target === 0) return 0.0
      return Math.round((target / achievement) * 10000) / 10000

    case 'timeline':
      if (!completionDate) return 0.0
      if (!deadline) return 1.0
      const comp = new Date(completionDate)
      const due = new Date(deadline)
      if (comp <= due) return 1.0
      const daysLate = Math.floor((comp.getTime() - due.getTime()) / (1000 * 60 * 60 * 24))
      return Math.max(0, 1.0 - daysLate / 30)

    case 'zero':
      return achievement === 0 ? 1.0 : 0.0

    default:
      return 0
  }
}

export function scoreToPercent(score: number): number {
  return Math.min(Math.round(score * 100 * 10) / 10, 150)
}

export function getScoreColor(pct: number): string {
  if (pct >= 100) return '#10b981' // emerald
  if (pct >= 80) return '#f59e0b'  // amber
  if (pct >= 50) return '#06b6d4'  // cyan
  return '#f43f5e'                  // rose
}

export function getUoMLabel(type: UoMType): string {
  const map: Record<UoMType, string> = {
    min: 'Min (↑ Higher is Better)',
    max: 'Max (↓ Lower is Better)',
    timeline: 'Timeline (Date-based)',
    zero: 'Zero-based (0 = 100%)',
  }
  return map[type] || type
}

export function getUoMDescription(type: UoMType): string {
  const map: Record<UoMType, string> = {
    min: 'e.g., Sales Revenue, Units Sold',
    max: 'e.g., TAT, Cost, Defect Rate',
    timeline: 'e.g., Project Delivery, Migration',
    zero: 'e.g., Safety Incidents, Critical Defects',
  }
  return map[type] || ''
}

File: backend/utils/scoring.py
from datetime import date
from typing import Optional


def compute_score(
    uom_type: str,
    target: float,
    achievement: float,
    completion_date: Optional[date] = None,
    deadline: Optional[date] = None,
) -> float:
    """
    Computes the achievement score for a goal based on its UoM type.

    Returns a float from 0.0 to 1.0 (may exceed 1.0 for overachievement).
    Multiply by 100 to get a percentage.

    uom_type options:
      min      - Higher actual is better (e.g. revenue, units sold). Score = actual / target.
      max      - Lower actual is better (e.g. TAT, cost). Score = target / actual.
      timeline - Binary pass/fail based on completion date vs deadline. Late completions are penalised.
      zero     - Zero incidents = 100%, anything else = 0%. Used for safety/compliance.
    """
    if uom_type == "min":
        if target == 0:
            return 1.0 if achievement >= 0 else 0.0
        return round(achievement / target, 4)

    elif uom_type == "max":
        if achievement == 0:
            return 1.0
        if target == 0:
            return 0.0
        return round(target / achievement, 4)

    elif uom_type == "timeline":
        if completion_date is None:
            return 0.0
        if deadline is None:
            return 1.0
        if completion_date <= deadline:
            return 1.0
        # Penalty: -3.33% per day late over a 30-day window, floor at 0
        days_late = (completion_date - deadline).days
        penalty = days_late / 30
        return round(max(0.0, 1.0 - penalty), 4)

    elif uom_type == "zero":
        return 1.0 if achievement == 0 else 0.0

    else:
        raise ValueError(f"Unknown UoM type: {uom_type}")


def score_to_percentage(score: float) -> float:
    """Converts a raw score to a display percentage, capped at 150 (overachievement ceiling)."""
    return round(min(score * 100, 150.0), 1)


def compute_weighted_score(goals_with_scores: list) -> float:
    """
    Computes the overall weighted average score across all goals.
    Input: list of (goal, score) tuples where goal.weightage is a numeric weight.
    """
    total_weight = sum(g.weightage for g, _ in goals_with_scores)
    if total_weight == 0:
        return 0.0
    weighted_sum = sum(score * g.weightage for g, score in goals_with_scores)
    return round(weighted_sum / total_weight, 4)
